src/natcap/invest/managed_raster/managed_raster.pyx

```diff
@@ -300,3 +309,119 @@ cdef class _ManagedRaster:
+            space), and ``flow_proportion`` (fraction of the flow
+            from the neighbor that flows to the original pixel).
+        """
+        cdef int n_dir, flow_dir_j, idx
+        cdef long xj, yj
+        cdef float flow_ji, flow_dir_j_sum
+
+        cdef NeighborTuple n
+        cdef vector[NeighborTuple] upslope_neighbor_tuples
+
+        for n_dir in range(8):
+            xj = xi + COL_OFFSETS[n_dir]
+            yj = yi + ROW_OFFSETS[n_dir]
+            if (xj < 0 or xj >= self.raster_x_size or
+                    yj < 0 or yj >= self.raster_y_size):
+                continue
+            flow_dir_j = <int>self.get(xj, yj)
+            flow_ji = (0xF & (flow_dir_j >> (4 * FLOW_DIR_REVERSE_DIRECTION[n_dir])))
+
+            if flow_ji:
+                flow_dir_j_sum = 0
+                for idx in range(8):
+                    flow_dir_j_sum += (flow_dir_j >> (idx * 4)) & 0xF
+                n.direction = n_dir
+                n.x = xj
+                n.y = yj
+                n.flow_proportion = flow_ji / flow_dir_j_sum
+                upslope_neighbor_tuples.push_back(n)
+
+        return upslope_neighbor_tuples
+
+    @cython.cdivision(True)
+    cdef vector[NeighborTuple] get_downslope_neighbors(
+            ManagedFlowDirRaster self, long xi, long yi, bint skip_oob=True):
+        """Return downslope neighbors of a given pixel.
+
+        Args:
+            xi (int): x coord in pixel space of the pixel to consider
+            yi (int): y coord in pixel space of the pixel to consider
+            skip_oob (bool): if True, do not return neighbors that fall
+                outside the raster bounds.
+
+        Returns:
+            libcpp.vector of NeighborTuples. Each NeighborTuple has
+            the attributes ``direction`` (integer flow direction 0-7
+            of the neighbor relative to the original pixel), ``x``
+            and ``y`` (integer coordinates of the neighbor in pixel
+            space), and ``flow_proportion`` (fraction of the flow
+            from the neighbor that flows to the original pixel).
+        """
+        cdef int n_dir
+        cdef long xj, yj
+        cdef float flow_ij
+
+        cdef NeighborTuple n
+        cdef vector[NeighborTuple] downslope_neighbor_tuples
+
+        cdef int flow_dir = <int>self.get(xi, yi)
+        cdef float flow_sum = 0
+
+        cdef int i = 0
+        for n_dir in range(8):
+            # flows in this direction
+            xj = xi + COL_OFFSETS[n_dir]
+            yj = yi + ROW_OFFSETS[n_dir]
+            if skip_oob and (xj < 0 or xj >= self.raster_x_size or
+                    yj < 0 or yj >= self.raster_y_size):
+                continue
+            flow_ij = (flow_dir >> (n_dir * 4)) & 0xF
+            flow_sum += flow_ij
+            if flow_ij:
+                n.direction = n_dir
+                n.x = xj
+                n.y = yj
+                n.flow_proportion = flow_ij
+                downslope_neighbor_tuples.push_back(n)
```

Comment:
Oh interesting. It must be creating copies somewhere, because it's returning the expected results. You're right it makes more sense to explicitly allocate the new object.